### ISSUE TITLE
PayFlow - Duplicate Charges..??

### ISSUE THREAD

**Issue by angelleye** [angelleye/paypal-woocommerce#454]
We've been getting reports lately from users experiencing duplicate orders getting processed through PayFlow.

One order succeeds, and the other at almost the same exact timestamp gets declined as a duplicate.  

We need to make sure the submit button is getting disabled when people click in, and check into any other reason these could be getting sent twice.

**Comment by angelleye** [comment]
I have 3 or 4 people reporting this same thing going on right now.  The transactions happen at the same time, but it's not just a double click of the submit button from what I'm seeing.  In each report it's happening on every single transaction.  For some reason the PayFlow call is happening twice.

So I'm getting it enough that it's worrisome, but I'm not getting it enough to think it's a global problem.

Not really sure how to troubleshoot this one, but we need to run some more tests in the PayFlow sandbox and then check our PayFlow sandbox account to make sure we're only seeing 1 transaction per order attempt.

**Comment by kcppdevelopers** [comment]
I have checked above issue, it seems working fine for me. if possible, please provide user website so I can move forward with it.

May be caching plugins or CDN enabled?

**Comment by angelleye** [comment]
I got some more info on this from PayPal.  Apparently it has something to do with the way we're seeing the X-VPS-REQUEST-ID header value.  See this documentation for more info on that:  https://developer.paypal.com/docs/classic/payflow/integration-guide/#payflow-header-parameters

The way we're setting that value in our code is: 

```
$unique_id = date('ymd-H').rand(1000,9999);
$headers[] = "X-VPS-Request-ID:" . $unique_id;
```

The way we're generating that must not be truly random..??  

I sent an example to PayPal from one of the user's reporting this and this is the response they said...

>You are getting a duplicate =1 error because you are using the same X-VPS-Request-ID:160428-018378 on PNREF B70P8FC4B193 as you did on PNREF B70P8FC4B190.

We need to look into this and make sure that value is always evaluating to a unique value for each unique transaction.

**Comment by angelleye** [comment]
Turns out the particular user having this issue was setup for PayFlow Link instead of PayFlow Pro.  

PayPal also provided this recommendation...

>I asked one of our developers about your PHP code.
>$unique_id = date('ymd-H').rand(1000,9999);

>He suggested pulling the system time to the nano second then throwing a random string on the end of it which would get it out to the billions.

**Comment by angelleye** [comment]
I'm re-opening this because as we discussed together there is still a bug here we need to get worked out.  You mentioned you had found it, so hopefully it won't be long before it's fixed..??

I'm also re-opening this to gather a collection of the people I need to report back to once we've merged this into the release branch.

**Comment by angelleye** [comment]
This has been fixed in #457
